sstables/trie/trie_writer.hh

```diff
@@ -0,0 +1,1192 @@
+// A payload consists of a 4-bit header and several bytes of content.
+struct trie_payload {
+    // What the reader will see is a 4-bit "payload bits" header, and a pointer to the body ("bytes") of the payload.
+    // The reader will use the header to determine the length and the meaning of the "bytes".
+    //
+    // The meaning of "bits" and "bytes" is generally opaque to the trie-writing layer,
+    // except that the value `bits == 0` must represent "no payload", and "bytes" is assumed
+    // empty in this case.
+    uint8_t _payload_bits = 0;
+    // We store the payload in a std::array + size pair,
+    // since it makes managing allocations slightly easier.
+    //
+    // 20 bytes is the biggest maximum possible payload in the BTI index format,
+    // so we don't need more.
+    std::array<std::byte, 20> _payload_buf = {};
+    // For the writer, the meaning of the "bits" is opaque,
+    // we just store an explicit size of the "bytes".
+    uint8_t _payload_size = 0;
+
+    trie_payload() noexcept = default;
+    trie_payload(uint8_t bits, const_bytes blob) noexcept {
+        expensive_assert(blob.size() <= _payload_buf.size());
+        expensive_assert(bits < 16);
+        expensive_assert(bool(blob.size()) == bool(bits));
+        _payload_bits = bits;
+        _payload_size = blob.size();
+        std::ranges::copy(blob, _payload_buf.data());
+    }
+    bool operator==(const trie_payload& other) const {
+        return other._payload_bits == _payload_bits
+            && std::ranges::equal(other.blob(), blob());
+    }
+
+    const_bytes blob() const noexcept{
+        return {_payload_buf.data(), _payload_size};
+    }
+};
+
+// The below few types are just strongly-typed wrappers around integers,
+// just so they can have some extra correctness checks and
+// so that we can write .valid() instead of "!= -1".
+// I'm not sure if this is better or worse than just using regular integers.
+
+// Represents a difference between positions in the output stream.
+// Just a strongly-typed wrapper around int64_t.
+struct sink_offset {
+    int64_t value = -1;
+    sink_offset() = default;
+    explicit sink_offset(int64_t v) : value(v) {}
+    std::strong_ordering operator<=>(const sink_offset& o) const {
+        expensive_assert(valid());
+        expensive_assert(o.valid());
+        return value <=> o.value;
+    };
+    bool operator==(const sink_offset& o) const = default;
+    bool valid() const {
+        return value >= 0;
+    }
+    sink_offset operator+(sink_offset o) const {
+        expensive_assert(valid());
+        expensive_assert(o.valid());
+        return sink_offset{value + o.value};
+    }
+};
+
+// Represents a size of a serialized node.
+// Just a strongly-typed wrapper around int16_t.
+struct node_size {
+    int16_t value = -1;
+    node_size() = default;
+    explicit node_size(int16_t v) : value(v) {}
+    bool valid() const {
+        return value >= 0;
+    }
+    operator sink_offset() const {
```

Comment:
Shouldn't it be explicit? implicit conversion undoes strong typing.

Reply:
Not in this case, this was deliberate. `node_size` is in some sense a subtype of `sink_offset`.